### PR TITLE
feat: application set handler

### PR DESCRIPTION
**Pull request by Greyeye**
# Change

1. **Application Set Check**:
    * as mentioned in [issue here](https://github.com/zapier/kubechecks/issues/227), a new changes how kubechecks handle the ApplicationSet. A change to spec.template.source will show the diff output for the matching Applications.
2. **kubechecks Helm Chart Version Up**: 
    * To accommodate the applicationSet check, kubechecks will need to access Kubernetes applicationSet CRD and security object to list the ArgoCD Clusters information. ClusterRole has been updated to grant these access.
3. **new kubechecks-rbac Helm Chart**: 
    * This helm chart Kubechekcs-rbac can be used to help grant access to argocd CRD and Clusters list from the remotely running Kubechecks. This is no required to be installed when Kubechecks is running on the same cluster with ArgoCD.
4. **AppWatcher Initialisation Change**:
    * Moved the AppWatcher's kubeclient initialisation to the container. rest.Config is passed as a parameter to AppWatcher.
5. **Fork of Generators Code from Argo CD**:
    * Forked generator code from Argo CD, This is because the kubechecks' go-gitlab v0.105 is not compatible with argo-cd's v0.91

# Local Dev Change
1. **Kubernetes Client Package**:
    * Implemented a new Kubernetes client specifically for AWS EKS.
    * Automated STS Authentication for EKS Client
    * Added a custom HTTP transport to the EKS client to automate STS token refresh.
2. **Extended Cobra Flags and Server Configuration**:
    * Added support for kubernetes-type (e.g., eks, local), kubernetes-clusterid, and kubernetes-cluster-region in the server configuration and CLI flags.

### CODE REVIEW

**Comment by Greyeye** [comment]
@djeebus @polyrain 

would you be kind enough to review the PR?


* You can ignore the pkg/generator directory as it's just a fork from Argo CD.  
* New Helm Chart: kubechecks-rbac:
    * I've added a new Helm chart named kubechecks-rbac to be installed where Argo CD is deployed.
    * This chart is intended to allow kubechecks-staging to access Argo CD's CRDs and Cluster list. I haven't had the opportunity to test it yet. If you have any concerns or suggestions for a better approach, I'd greatly appreciate your input.

* Architecture Documentation:
    * I've included an event flow diagram in the architecture documentation. It would be great if you could review this as well.

**Review by MeNsaaH — CHANGES_REQUESTED** [review]
Some code changes and small discussions. As kubechecks grows, it becomes paramount to add tests for code modifications 😄 , hence why you see me suggesting on tests. 

Thank you for your contribution

File: localdev/terraform/modules/vcs_files/base_files/appsets/echo-server/.gitkeep
[the remark this answers is not in the document]
Comment (Greyeye) [reply localdev/terraform/modules/vcs_files/base_files/appsets/echo-server/.gitkeep:1]: it doesnt get auto created, but i made changed to the command.

File: localdev/test_appsets/Tiltfile [hunk 1]
```diff
@@ -3,28 +3,20 @@
 # Test ArgoCD Applications
 # /////////////////////////////////////////////////////////////////////////////
 
-def install_test_appsets(cfg):
+def copy_test_appsets(cfg):
```
[the remark this answers is not in the document]
Comment (Greyeye) [reply localdev/test_appsets/Tiltfile:6]: appsets are handled by app-in-app file ./localdev/test_apps/app-root.yaml

File: pkg/app_watcher/appset_watcher.go
[the remark this answers is not in the document]
Comment (Greyeye) [reply pkg/app_watcher/appset_watcher.go:1]: sure will add one tomorrow

File: pkg/appdir/app_directory.go [hunk 2]
```diff
@@ -58,13 +59,16 @@ func (d *AppDirectory) ProcessApp(app v1alpha1.Application) {
 	}
 }
 
-func (d *AppDirectory) FindAppsBasedOnChangeList(changeList []string, targetBranch string) []v1alpha1.Application {
+// FindAppsBasedOnChangeList receives the modified file path and
+// returns the list of applications that are affected by the changes.
+//
+//	e.g. changeList = ["path/to/file1", "path/to/file2"]
+func (d *AppDirectory) FindAppsBasedOnChangeList(changeList []string, targetBranch string, _ *git.Repo) []v1alpha1.Application {
```
[the remark this answers is not in the document]
Comment (Greyeye) [reply pkg/appdir/app_directory.go:66]: good point, updated it.

File: charts/kubechecks-rbac/templates/role.yaml [hunk 3]
```diff
@@ -0,0 +1,11 @@
+apiVersion: rbac.authorization.k8s.io/v1
+kind: Role
+metadata:
+  name: {{ .Values.roleName | default "kubechecks-remote-role" }}
```
[the remark this answers is not in the document]
Comment (MeNsaaH) [reply charts/kubechecks-rbac/templates/role.yaml:4]: IMO, kubechecks staging should be deployed to tooling EKS then. This increases the complexity of kubechecks because now, we have to worry about external cluster dependencies, authentication and refreshing credentials 🤔 . @djeebus  what do you think?

File: pkg/container/main.go [hunk 4]
```diff
@@ -16,23 +17,28 @@ import (
...
 
 	VcsClient    vcs.Client
 	VcsToArgoMap VcsToArgoMap
+
+	KubeClientSet client.Interface
 }
 
 type VcsToArgoMap interface {
 	AddApp(*v1alpha1.Application)
+	AddAppSet(*v1alpha1.ApplicationSet)
```
[the remark this answers is not in the document]
Comment (Greyeye) [reply pkg/container/main.go:36]: good point, fixed it

File: localdev/test_appsets/Tiltfile (hunk 1 above)
[the remark this answers is not in the document]
Comment (MeNsaaH) [reply localdev/test_appsets/Tiltfile:6]: That will then just create a new files untracked by git at `/localdev/terraform/modules/vcs_files/base_files/appsets/`, doesn't it?

File: pkg/events/check.go [hunk 5]
```diff
@@ -269,7 +282,7 @@ func (ce *CheckEvent) Process(ctx context.Context) error {
 
 	close(ce.appChannel)
 
-	ce.logger.Debug().Msg("finished an app")
+	ce.logger.Debug().Stack().Msg("finished an app/appsets")
```
[the remark this answers is not in the document]
Comment (Greyeye) [reply pkg/events/check.go:285]: my bad, Stack() is used with err() and can show stack trace. it does not work with Debug()
removed it.

File: pkg/generator/applicationsets.go
[the remark this answers is not in the document]
Comment (Greyeye) [reply pkg/generator/applicationsets.go:1]: problem is with this [code](https://github.com/zapier/kubechecks/blob/49fc58db846ac10232d2bd80d4d0990df2364fb4/pkg/vcs/gitlab_client/merge.go#L32)
it's a new function that isn't available on v0.91, and v0.105 breaks the code inside argocd, and prevented it from building.
I dont have a good solution for this unless @djeebus got better idea?

File: localdev/test_appsets/Tiltfile (hunk 1 above)
[the remark this answers is not in the document]
Comment (Greyeye) [reply localdev/test_appsets/Tiltfile:6]: thats right, these files are added to gitignore already

File: pkg/appdir/vcstoargomap.go [hunk 6]
```diff
@@ -98,6 +114,41 @@ func (v2a VcsToArgoMap) GetVcsRepos() []string {
...
 		repos = append(repos, key.CloneURL(v2a.username))
 	}
-
+	for key := range v2a.appSetDirByRepo {
+		repos = append(repos, key.CloneURL(v2a.username))
+	}
 	return repos
 }
+
+func (v2a VcsToArgoMap) AddAppSet(app *v1alpha1.ApplicationSet) {
```
[the remark this answers is not in the document]
Comment (Greyeye) [reply pkg/appdir/vcstoargomap.go:123]: added tests and renamed the file. 
vcs to argu ? map??

File: pkg/events/check.go [hunk 7]
```diff
@@ -110,11 +111,23 @@ func (ce *CheckEvent) GenerateListOfAffectedApps(ctx context.Context, repo *git.
...
 		ce.logger.Error().Err(err).Msg("could not get list of affected apps and appsets")
 	}
+
+	for _, appSet := range ce.affectedItems.ApplicationSets {
+		apps, err := generator.GenerateApplicationSetApps(ctx, appSet, &ce.ctr)
+		if err != nil {
+			ce.logger.Error().Err(err).Msg("could not generate apps from appSet")
+			continue
+		}
+		ce.affectedItems.Applications = append(ce.affectedItems.Applications, apps...)
```
[the remark this answers is not in the document]
Comment (djeebus) [reply pkg/events/check.go:128]: On the one hand, we'll end up with duplicates in this slice. On the other hand, the `CheckEvent.queueApp` function (line 314 of the same file) will avoid diffing dupes (around lines 321-325). We should be fine, and I don't think we need to spend the extra effort to avoid the dupes in this slice. Good call out though!

[the remark this answers is not in the document]
Comment (djeebus) [reply pkg/events/check.go:128]: (also :+1: on the love for tests!)

File: cmd/root.go [hunk 8]
```diff
@@ -63,6 +63,12 @@ func init() {
 	stringFlag(flags, "argocd-api-namespace", "ArgoCD namespace where the application watcher will read Custom Resource Definitions (CRD) for Application and ApplicationSet resources.",
 		newStringOpts().
 			withDefault("argocd"))
+	stringFlag(flags, "kubernetes-type", "Kubernetes Type One of eks, or localhost. Defaults to localhost.",
+		newStringOpts().
+			withChoices("eks", "localhost").
+			withDefault("localhost"))
```
Comment (djeebus) [inline cmd/root.go:69]: "localhost"? "local" might make more sense, since "localhost" normally refers to an ip address.

File: charts/kubechecks-rbac/templates/role.yaml (hunk 3 above)
[the remark this answers is not in the document]
Comment (djeebus) [reply charts/kubechecks-rbac/templates/role.yaml:4]: In the future, I think we do want to support the idea of hosting kubechecks outside of the cluster that argocd is running in. I'm not opposed to this change, mostly because this is opt-in behavior, and if you do opt-in, it's going to be complicated one way or another.

File: pkg/generator/applicationsets.go
[the remark this answers is not in the document]
Comment (djeebus) [reply pkg/generator/applicationsets.go:1]: Nope, no better options here. This code really isn't intended to be used in the way we're using it, so either we fork or give up, and I'd rather not give up. It would be nice to get a doc in these folders indicating where the code was copied from and why.

[the remark this answers is not in the document]
Comment (Greyeye) [reply pkg/generator/applicationsets.go:1]: i'll add readme.md 
might raise a PR to argo-cd to get their go-gitlab updated but thats gonna be a long way off.

File: cmd/root.go (hunk 8 above)
Comment (Greyeye) [reply cmd/root.go]: changed to `local`

File: pkg/generator/applicationsets.go
[the remark this answers is not in the document]
Comment (Greyeye) [reply pkg/generator/applicationsets.go:1]: added README.md

File: pkg/events/check.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (Greyeye) [reply pkg/events/check.go:139]: I had to refactor Check.go 's GenerateListOfAffectedApps quite a bit to make it testable. Added the test just now.

File: pkg/app_watcher/appset_watcher.go
[the remark this answers is not in the document]
Comment (Greyeye) [reply pkg/app_watcher/appset_watcher.go:1]: this has been added

File: charts/kubechecks-rbac/templates/role.yaml (hunk 3 above)
[the remark this answers is not in the document]
Comment (MeNsaaH) [reply charts/kubechecks-rbac/templates/role.yaml:4]: It's definitely a nice to have. I'm just looking at the complexity it'll require to setup. It's a gate that once opened will flood in lots of requirements for different cluster setups, eks with OIDC, eks with rancher setup etc. If we can avoid inter-cluster auth, I think we should 😄. I'm just being too careful to make sure we are not the ones to open pandoras box ourself. Maybe allow for folks to request it?

[the remark this answers is not in the document]
Comment (Greyeye) [reply charts/kubechecks-rbac/templates/role.yaml:4]: The /pkg/kubernetes package currently supports both EKS and in-cluster setups. To call the backend API for EKS, the caller must have an assigned IAM role. This is a straightforward use case. (sysadmin can assign role using kube2iam/IRSA, or even use env var to get the IAM role)

Regarding OIDC or Rancher, they do not impact the current functionality described above.

In the future, I was hoping to add support for Azure Kubernetes Service (AKS) and Google Kubernetes Engine (GKE).

File: cmd/container.go [hunk 9]
```diff
@@ -36,7 +37,30 @@ func newContainer(ctx context.Context, cfg config.ServerConfig, watchApps bool)
...
+	// TODO: expand with other cluster types
+	case client.ClusterTypeLOCAL:
+		kubeClient, err = client.New(&client.NewClientInput{
+			KubernetesConfigPath: cfg.KubernetesConfig,
+			ClusterType:          cfg.KubernetesType,
+		})
+		if err != nil {
+			return ctr, errors.Wrap(err, "failed to create kube client")
+		}
+	case client.ClusterTypeEKS:
```
[the remark this answers is not in the document]
Comment (djeebus) [reply cmd/container.go:52]: The alternative is pretty gnarly. IAM roles, cross account trusts, etc, assuming it can actually be done. This seems relatively simple (compared to the rest of the features), and fairly isolated/repeatable.

File: pkg/kubernetes/api_eks_client.go [hunk 10]
```diff
@@ -0,0 +1,188 @@
...
+// accessing the EKS cluster.
+//
+// Parameters:
+//
+//   - ctx: the context for the AWS SDK client.
+//
+//   - clusterID: the name of the EKS cluster.
+//
+//   - region: the AWS region where the EKS cluster is located. (optional) If not provided, the default region will be used.
+func EKSClientOption(ctx context.Context, clusterID, region string) NewClientOption {
```
Comment (djeebus) [inline pkg/kubernetes/api_eks_client.go:71]: I don't think you need `region`, since `AWS_REGION` will let users override this themselves directly.

File: pkg/appdir/app_directory.go [hunk 11]
```diff
@@ -146,10 +150,15 @@ func (d *AppDirectory) GetApps(filter func(stub v1alpha1.Application) bool) []v1
 }
 
 func (d *AppDirectory) AddApp(app v1alpha1.Application) {
+	if _, exists := d.appsMap[app.Name]; exists {
+		log.Info().Msgf("app %s already exists", app.Name)
```
[the remark this answers is not in the document]
Comment (djeebus) [reply pkg/appdir/app_directory.go:153]: It does, but if an app with the same name is added multiple times, we're already in a weird state. I don't think argocd can handle that scenario, so avoiding extra work is _probably_ a reasonable path to take in this undefined state. It's definitely an edge case one way or the other though.

**Comment by djeebus** [comment]
I saw your successful test; nice work! I've also confirmed backwards compatibility. :+1:

**Review by djeebus — APPROVED** [review]

File: pkg/kubernetes/api_eks_client.go (hunk 10 above)
Comment (Greyeye) [reply pkg/kubernetes/api_eks_client.go]: @djeebus i've removed the region parameter.

File: charts/kubechecks-rbac/values.yaml [hunk 12]
```diff
@@ -1,13 +1,6 @@
...
-roleBindingName: "kubechecks-remote-role-binding"
-namespace: "argocd"
-
-# for AWS EKS clusters
-eks:
-  # The ARN of the IAM role to bind to the Role
-  iamRoleArn: ""
+ClusterRoleName: "kubechecks-remote-role"
+ClusterRoleBindingName: "kubechecks-remote-role-binding"
+ClusterRoleBindingGroup: "kubechecks-remote-group"
```
Comment (djeebus) [inline charts/kubechecks-rbac/values.yaml:3]: these should start with a lowercase letter, for consistency

File: pkg/argo_client/applications.go [hunk 13]
```diff
@@ -132,8 +132,8 @@ func (argo *ArgoClient) GetApplicationSets(ctx context.Context) (*v1alpha1.Appli
 
 	resp, err := appClient.List(ctx, new(applicationset.ApplicationSetListQuery))
 	if err != nil {
-		telemetry.SetError(span, err, "Argo List All Applications error")
-		return nil, errors.Wrap(err, "failed to applications")
+		telemetry.SetError(span, err, "Argo List All Application Sets error")
+		return nil, errors.Wrap(err, "failed to application sets")
```
Comment (djeebus) [inline pkg/argo_client/applications.go:136]: probably should be "failed to list application sets"

File: charts/kubechecks-rbac/values.yaml (hunk 12 above)
Comment (Greyeye) [reply charts/kubechecks-rbac/values.yaml]: good point, changed to lowercase

File: pkg/argo_client/applications.go (hunk 13 above)
Comment (Greyeye) [reply pkg/argo_client/applications.go]: changed it to failed to list application sets.

**Review by MeNsaaH — APPROVED** [review]
